scripts/finalizeCCTPMessages.ts

```diff
@@ -0,0 +1,309 @@
+      },
+    ],
+    name: "receiveMessage",
+    outputs: [
+      {
+        internalType: "bool",
+        name: "success",
+        type: "bool",
+      },
+    ],
+    stateMutability: "nonpayable",
+    type: "function",
+  },
+  {
+    inputs: [
+      {
+        internalType: "bytes32",
+        name: "",
+        type: "bytes32",
+      },
+    ],
+    name: "usedNonces",
+    outputs: [
+      {
+        internalType: "uint256",
+        name: "",
+        type: "uint256",
+      },
+    ],
+    stateMutability: "view",
+    type: "function",
+  },
+];
+
+/**
+ * Script to claim L1->L2 or L2->L1 messages. Run via
+ * ```
+ * RECEIVE_MESSAGES_ON=l1 \
+ * yarn hardhat run ./scripts/claimLineaMessages.ts \
+ * --network linea-goerli \
+ * ```
+ * Environment variables:
+ * - `RECEIVE_MESSAGES_ON`: Which messages to claim. Either `l1` or `l2`.
+ * Flags:
+ * - `--network`: The L2 network, i.e. `linea-goerli` or `linea`.
+ *
+ */
+async function main() {
+  const receiveMessagesOn = process.env.RECEIVE_MESSAGES_ON || "l2";
+  const l1BlockLookback = parseInt(process.env.BLOCK_LOOKBACK || "8640");
+  const l2BlockLookback = parseInt(process.env.BLOCK_LOOKBACK || "17380");
+
+  const l1ChainId = parseInt(await hre.companionNetworks.l1.getChainId());
+  const l2ChainId = parseInt(await hre.getChainId());
+  const l1Provider = new ethers.providers.JsonRpcProvider(
+    getNodeUrl(l1ChainId === 1 ? "mainnet" : "goerli", true, l1ChainId)
+  );
+  const l2Provider = ethers.provider;
+  const l1Signer = ethers.Wallet.fromMnemonic((hre.network.config.accounts as any).mnemonic).connect(l1Provider);
+  const [l2Signer] = await ethers.getSigners();
+
+  console.log("\nL1 chain ID:", l1ChainId);
+  console.log("L2 chain ID:", l2ChainId);
+  console.log("Signer address:", l1Signer.address);
+
+  // Get relevant src events
+  const srcEvents =
+    receiveMessagesOn === "l1"
+      ? await getL2SrcEvents(l2ChainId, l2Provider, l2Signer, l2BlockLookback)
+      : await getL1SrcEvents(l1ChainId, l1Provider, l1Signer, l1BlockLookback);
+
+  // Get relevant message hashes and bytes
+  const uniqueTxHashes = new Set(srcEvents.map((event) => event.transactionHash));
+  const messageHashesAndBytes = await parseMessageHashesAndBytes(
+    Array.from(uniqueTxHashes),
+    receiveMessagesOn === "l1" ? l2Provider : l1Provider
+  );
+  const relevantMessageHashesAndBytes = messageHashesAndBytes.filter(
+    (messageHashAndBytes) =>
+      messageHashAndBytes.destinationDomain === CIRCLE_DOMAIN_IDs[receiveMessagesOn === "l1" ? l1ChainId : l2ChainId]
+  );
+  console.log(`\nParsed ${relevantMessageHashesAndBytes.length} relevant 'MessageSent' events`);
+
+  if (relevantMessageHashesAndBytes.length === 0) {
+    console.log("No relevant messages to receive");
+    return;
+  }
+
+  // Request attestations
+  console.log("\nRequesting attestations...");
+  const attestations = [];
+  for (const messageHashAndBytes of relevantMessageHashesAndBytes) {
```

Comment:
Should we send these requests in parallel?

Reply:
The Circle API used for attestations is rate-limited to 10 req/sec. I wanted to have something working without implementing batch request logic. Also, it was easier to implement sequential logic here for debugging purposes. I think for the scope of this script, this suffices but in the context of the dataworker it would make sense to make these requests in parallel.

Reply:
I agree -- I think when we migrate this to a bot, we would want to optimize it, but here it's okay if it's slow.